### PR TITLE
[cxx-interop] Ban ObjCBool from being substituted into C++ templates

### PR DESCRIPTION
**Pull request by egorzhdan**
This fixes a compiler crash when calling a templated C++ function with a parameter of type `ObjCBool` from Swift. The compiler now emits an error for this.

Previously the following would happen:
1. Swift starts to emit SILGen for a call expression `takeTAsConstRef(ObjCBool(true))`.
2. `takeTAsConstRef` is a templated C++ function, so Swift asks Clang to instantiate a `takeTAsConstRef` with a built-in Obj-C boolean type.
3. Swift gets an instantiated function template back that looks like this: `void takeTAsConstRef(_Bool t) { ... }`.
4. Swift's ClangImporter begins to import that instantiated function.
5. Since the parameter type is not spelled as `BOOL`, the parameter is not imported as `ObjCBool`. Instead, it's imported as regular Swift `Bool`.
6. Swift realizes that the types don't match (`ObjCBool` vs `Bool`), tries to apply any of the known implicit conversions, fails to do so, crashes.

rdar://130424969

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by egorzhdan** [comment]
@swift-ci please test

File: test/Interop/Cxx/templates/function-template-objc-typechecker.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+// REQUIRES: objc_interop
+
+import FunctionTemplates
+import ObjectiveC
+
+// Verify that ObjCBool type is banned from C++ template parameters.
+
+takesValue(ObjCBool(true))
+// CHECK: error: could not generate C++ types from the generic Swift types provided; the following Swift type(s) provided to 'takesValue' could not be converted: ObjCBool
```
Comment (ravikandhadai) [inline test/Interop/Cxx/templates/function-template-objc-typechecker.swift:11]: Is there a way to customize this error message for this case? Can we emit a fix-it that says use `.boolValue` to convert to a Swift boolean?

**Review by beccadax — APPROVED** [review]
Question. I notice this line in `ClangTypeConverter::visitStructType()` (ClangTypeConverter.cpp:265 or so):

```cpp
  CHECK_NAMED_TYPE("ObjCBool", ctx.ObjCBuiltinBoolTy);
```

If you change `ctx.ObjCBuiltinBoolTy` to `ctx.getBOOLType()`, do these templates start importing correctly? I think that might help explain why `ObjCBool` isn't round-tripping.

If that *doesn't* help, this PR looks like a good band-aid to prevent the crash, but it seems like a better fix ought to be possible in the future.

**Comment by egorzhdan** [comment]
> If you change ctx.ObjCBuiltinBoolTy to ctx.getBOOLType(), do these templates start importing correctly? I think that might help explain why ObjCBool isn't round-tripping.

That's a good question. I initially tried that, and found out that `ctx.getBOOLType()` returns nullptr until Clang's Sema processes an Obj-C `BOOL` literal, and then does a name lookup for `BOOL`. I tried duplicating that logic in this patch, but unfortunately the name lookup won't find `BOOL` unless `objc.h` is transitively included in the C++ module which is being processed.

File: test/Interop/Cxx/templates/function-template-objc-typechecker.swift (hunk 1 above)
Comment (egorzhdan) [reply test/Interop/Cxx/templates/function-template-objc-typechecker.swift:11]: Emitting a fix-it seems like a bit like an overkill for this case, I think. We would also need to handle possible explicit generic parameters (e.g. `myCxxMethod<T1, ObjCBool, T2>`). 
This bit of logic also doesn't have access to source locations. We could certainly pipe them through, but I'm not sure if it's worth the effort.

**Pull request merged**